### PR TITLE
[NOD-1423] Refactor the miner and mempool

### PR DESCRIPTION
**Pull request by elichai**
This is based on #972

### CODE REVIEW

File: domain/miningmanager/mempool/mempool.go [hunk 1]
```diff
@@ -1,36 +1,915 @@
...
+func (mp *mempool) ChainedCount() int {
+	mp.mtx.RLock()
+	defer mp.mtx.RUnlock()
+	return len(mp.chainedTransactions)
+}
+
+// MiningDescs returns a slice of mining descriptors for all the transactions
+// in the pool.
+//
+// This is part of the mining.TxSource interface implementation and is safe for
```
Comment (svarogg) [inline domain/miningmanager/mempool/mempool.go:854]: It seems like mining.TxSource is no longer a thing, if so - remove this comment (except the part about concurrent use)

File: domain/miningmanager/mempool/mempool.go [hunk 2]
```diff
@@ -1,36 +1,915 @@
...
+	i := 0
+	for _, desc := range mp.pool {
+		descs[i] = desc.DomainTransaction
+		i++
+	}
+
+	return descs
+}
+
+// HandleNewBlock removes all the transactions in the new block
```
Comment (svarogg) [inline domain/miningmanager/mempool/mempool.go:869]: HandleNewBlock -> HandleNewBlockTransactions

**Review by svarogg — CHANGES_REQUESTED** [review]

File: domain/miningmanager/blocktemplatebuilder/blocktemplatebuilder.go [hunk 3]
```diff
@@ -3,24 +3,171 @@ package blocktemplatebuilder
...
+func (btb *blockTemplateBuilder) calcTxValue(tx *consensusexternalapi.DomainTransaction) float64 {
+	massLimit := btb.policy.BlockMaxMass
+
+	mass := tx.Mass
+	fee := tx.Fee
+	if subnetworks.IsBuiltInOrNative(tx.SubnetworkID) {
+		return float64(fee) / (float64(mass) / float64(massLimit))
+	}
+
+	panic("We currently don't support non native subnetworks")
```
[the remark this answers is not in the document]
Comment (elichai) [reply domain/miningmanager/blocktemplatebuilder/blocktemplatebuilder.go:166]: The problem is that there's no `dag.GasLimit()` function anywhere

File: domain/miningmanager/mempool/mempool.go [hunk 4]
```diff
@@ -1,36 +1,915 @@
...
+// orphanTx is normal transaction that references an ancestor transaction
+// that is not yet available. It also contains additional information related
+// to it such as an expiration time to help prevent caching the orphan forever.
+type orphanTx struct {
+	tx         *consensusexternalapi.DomainTransaction
+	expiration mstime.Time
+}
+
+// Ensure the mempool type implements the mining.TxSource interface.
+//var _ mining.TxSource = (*mempool)(nil)
```
[the remark this answers is not in the document]
Comment (elichai) [reply domain/miningmanager/mempool/mempool.go:131]: Ops, was supposed to delete this

File: domain/miningmanager/mempool/mempool.go [hunk 5]
```diff
@@ -1,36 +1,915 @@
...
+// such as rejecting duplicate transactions, ensuring transactions follow all
+// rules, detecting orphan transactions, and insertion into the memory pool.
+//
+// If the transaction is an orphan (missing parent transactions), the
+// transaction is NOT added to the orphan pool, but each unknown referenced
+// parent is returned. Use ProcessTransaction instead if new orphans should
+// be added to the orphan pool.
+//
+// This function MUST be called with the mempool lock held (for writes).
+func (mp *mempool) maybeAcceptTransaction(tx *consensusexternalapi.DomainTransaction, rejectDupOrphans bool) ([]consensusexternalapi.DomainOutpoint, *txDesc, error) {
```
[the remark this answers is not in the document]
Comment (elichai) [reply domain/miningmanager/mempool/mempool.go:545]: It's a consensus rule, so in the consensus (via the `consensus.ValidateTransactionAndPopulateWithConsensusData(tx)` call)

File: domain/miningmanager/mempool/mempool.go [hunk 6]
```diff
@@ -1,36 +1,915 @@
...
+			if !found {
+				rejectCode = RejectNonstandard
+			}
+			str := fmt.Sprintf("transaction %s has a non-standard "+
+				"input: %s", txID, err)
+			return nil, nil, txRuleError(rejectCode, str)
+		}
+	}
+
+	// TODO: How to get fee/mass?
```
[the remark this answers is not in the document]
Comment (elichai) [reply domain/miningmanager/mempool/mempool.go:635]: Yep, it's a TODO I forgot to delete, thanks!

File: domain/miningmanager/mempool/error.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	if err == nil {
+		return RejectInvalid, false
+	}
+
+	return RejectInvalid, false
+}
+
+// ErrToRejectErr examines the underlying type of the error and returns a reject
+// code and string appropriate to be sent in a appmessage.MsgReject message.
+func ErrToRejectErr(err error) (RejectCode, string) {
```
[the remark this answers is not in the document]
Comment (elichai) [reply domain/miningmanager/mempool/error.go:154]: It's from the old mempool, and it looks like nothing ever called this? I guess I'll delete (there's even no `RejectErr` type)

File: domain/miningmanager/blocktemplatebuilder/blocktemplatebuilder.go [hunk 8]
```diff
@@ -3,24 +3,171 @@ package blocktemplatebuilder
...
+//  |  Low-fee/Non high-priority (free) |   |
+//  |  transactions (while block size   |   |
+//  |  <= policy.BlockMinSize)          |   |
+//   -----------------------------------  --
+
 func (btb *blockTemplateBuilder) GetBlockTemplate(coinbaseData *consensusexternalapi.DomainCoinbaseData) *consensusexternalapi.DomainBlock {
-	return nil
+	mempoolTransactions := btb.mempool.Transactions()
+	candidateTxs := make([]*candidateTx, 0, len(mempoolTransactions))
+	for _, tx := range mempoolTransactions {
```
[the remark this answers is not in the document]
Comment (elichai) [reply domain/miningmanager/blocktemplatebuilder/blocktemplatebuilder.go:109]: That's how it was before(See the `func (g *BlkTmplGenerator) collectCandidatesTxs` function), I'd rather minimize the modifications for the mempool in this PR

File: domain/miningmanager/blocktemplatebuilder/blocktemplatebuilder.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/miningmanager/blocktemplatebuilder/blocktemplatebuilder.go:166]: How about `gasLimit = math.MaxFloat64` and add TODO comment to replace it with real gas limit when it is implemented?

File: domain/miningmanager/mempool/error.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/miningmanager/mempool/error.go]: Yeah, please delete it.

File: domain/miningmanager/mempool/mempool.go [hunk 9]
```diff
@@ -1,36 +1,915 @@
...
+
+// addTransaction adds the passed transaction to the memory pool. It should
+// not be called directly as it doesn't perform any validation. This is a
+// helper for maybeAcceptTransaction.
+//
+// This function MUST be called with the mempool lock held (for writes).
+func (mp *mempool) addTransaction(tx *consensusexternalapi.DomainTransaction, mass uint64, fee uint64, parentsInPool []consensusexternalapi.DomainOutpoint) (*txDesc, error) {
+	// Add the transaction to the pool and mark the referenced outpoints
+	// as spent by the pool.
+	txD := &txDesc{
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/miningmanager/mempool/mempool.go]: This one still not resolved

File: domain/miningmanager/mempool/mempool.go [hunk 10]
```diff
@@ -632,7 +621,6 @@ func (mp *mempool) maybeAcceptTransaction(tx *consensusexternalapi.DomainTransac
 		}
 	}
 
-	// TODO: How to get fee/mass?
```
Comment (svarogg) [inline domain/miningmanager/mempool/mempool.go]: You also need to uncomment and modify the code below : `if txFee == 0....`

File: domain/miningmanager/mempool/mempool.go [hunk 11]
```diff
@@ -1,36 +1,900 @@
...
+
+// addTransaction adds the passed transaction to the memory pool. It should
+// not be called directly as it doesn't perform any validation. This is a
+// helper for maybeAcceptTransaction.
+//
+// This function MUST be called with the mempool lock held (for writes).
+func (mp *mempool) addTransaction(tx *consensusexternalapi.DomainTransaction, mass uint64, fee uint64, parentsInPool []consensusexternalapi.DomainOutpoint) (*txDescriptor, error) {
+	// Add the transaction to the pool and mark the referenced outpoints
+	// as spent by the pool.
+	txD := &txDescriptor{
```
Comment (svarogg) [inline domain/miningmanager/mempool/mempool.go:477]: I actually didn't mind txDesc too much (well, I did, but no commenting on names and everything), and what really bothered me was txD.

I didn't know that tx's had a D :stuck_out_tongue_closed_eyes: . Sorry not sorry.

**Review by svarogg — CHANGES_REQUESTED** [review]

File: domain/miningmanager/mempool/mempool.go (hunk 11 above)
Comment (elichai) [reply domain/miningmanager/mempool/mempool.go:477]: looool

File: domain/miningmanager/mempool/mempool.go (hunk 10 above)
Comment (elichai) [reply domain/miningmanager/mempool/mempool.go]: This is covered by the minRelayFee, will remove it

**Review by svarogg — APPROVED** [review]

**Pull request merged**